File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/subscribe/internal/pipeline/PcmRing.kt
package com.swmansion.moqkit.subscribe.internal.pipeline

import kotlin.math.ceil
import kotlin.math.min

internal data class PcmWriteResult(
    val acceptedFrames: Int,
    val rejectedOldFrames: Int = 0,
    val evictedFrames: Int = 0,
    val silenceFrames: Int = 0,
)

/** Timestamp-positioned PCM16 ring with explicit admission effects. */
internal class PcmRing(
    val sampleRate: Int,
    val channels: Int,
    policy: AdmissionPolicy,
) {
    private var samples = ShortArray(capacityFrom(policy) * channels)
    private var writeFrame = 0L
    private var readFrame = 0L
    private var anchored = false
    private var buffering = true

    init {
        require(sampleRate > 0) { "invalid sample rate" }
        require(channels > 0) { "invalid channels" }
    }

    val capacity: Int
        get() = samples.size / channels

    val length: Int
        get() = (writeFrame - readFrame).coerceIn(0L, capacity.toLong()).toInt()

    val timestampUs: Long
        get() = multiplyOrMax(readFrame, MICROS_PER_SECOND) / sampleRate

    fun write(timestampUs: Long, samples: ShortArray, frameCount: Int): PcmWriteResult {
        require(timestampUs >= 0L) { "timestampUs must be non-negative" }
        require(frameCount >= 0) { "frameCount must be non-negative" }
        require(samples.size >= frameCount * channels) { "insufficient PCM samples" }
        if (frameCount == 0) return PcmWriteResult(acceptedFrames = 0)

        var start = timestampToFrame(timestampUs)
        var frames = frameCount
        var sourceOffset = 0
        var rejectedOld = 0
        var evicted = 0

        if (!anchored) {
            anchored = true
            readFrame = start
            writeFrame = start
        }

        val oldFrames = (readFrame - start).coerceAtLeast(0L).coerceAtMost(frames.toLong()).toInt()
        if (oldFrames > 0) {
            rejectedOld += oldFrames
            start += oldFrames
            sourceOffset += oldFrames
            frames -= oldFrames
        }
        if (frames == 0) {
            return PcmWriteResult(acceptedFrames = 0, rejectedOldFrames = rejectedOld)
        }

        if (frames > capacity) {
            val trim = frames - capacity
            start += trim
            sourceOffset += trim
            frames -= trim
            evicted += trim
        }

        val end = addOrMax(start, frames.toLong())
        val overflow = (end - readFrame - capacity).coerceAtLeast(0L).toInt()
        if (overflow > 0) {
            readFrame += overflow
            evicted += overflow
        }

        val silence = if (start > writeFrame) {
            min(start - writeFrame, capacity.toLong()).toInt().also {
                ringZero(writeFrame, it)
            }
        } else {
            0
        }

        ringCopy(start, samples, sourceOffset, frames)
        if (end > writeFrame) writeFrame = end
        if (writeFrame - readFrame >= capacity) buffering = false

        return PcmWriteResult(
            acceptedFrames = frames,
            rejectedOldFrames = rejectedOld,
            evictedFrames = evicted,
            silenceFrames = silence,
        )
    }

    fun read(output: ShortArray, maxFrames: Int): Int {
        require(maxFrames >= 0) { "maxFrames must be non-negative" }
        require(output.size >= maxFrames * channels) { "insufficient output samples" }
        if (buffering || maxFrames == 0) return 0
        val frames = min(writeFrame - readFrame, maxFrames.toLong()).toInt()
        if (frames <= 0) return 0
        ringRead(readFrame, output, frames)
        readFrame += frames
        return frames
    }

    fun resize(policy: AdmissionPolicy) {
        val newCapacity = capacityFrom(policy)
        if (newCapacity == capacity) return

        val framesToKeep = min(length, newCapacity)
        val recent = ShortArray(framesToKeep * channels)
        if (framesToKeep > 0) ringRead(writeFrame - framesToKeep, recent, framesToKeep)
        samples = ShortArray(newCapacity * channels)
        readFrame = writeFrame - framesToKeep
        if (framesToKeep > 0) ringCopy(readFrame, recent, 0, framesToKeep)
        buffering = true
    }

    fun reset() {
        samples.fill(0)
        writeFrame = 0L
        readFrame = 0L
        anchored = false
        buffering = true
    }

    private fun capacityFrom(policy: AdmissionPolicy): Int {
        require(sampleRate > 0) { "invalid sample rate" }
        require(channels > 0) { "invalid channels" }
        val durationFrames = ceil(sampleRate.toDouble() * policy.maxDurationUs / MICROS_PER_SECOND)
            .toLong()
            .coerceAtLeast(1L)
        val byteFrames = policy.maxBytes / (channels * PCM_BYTES_PER_SAMPLE)
        return minOf(policy.maxFrames.toLong(), durationFrames, byteFrames)
            .coerceIn(1L, Int.MAX_VALUE.toLong())
            .toInt()
    }

    private fun timestampToFrame(timestampUs: Long): Long =
        Math.round(timestampUs.toDouble() * sampleRate / MICROS_PER_SECOND)

    private fun ringCopy(position: Long, source: ShortArray, sourceOffset: Int, frames: Int) {
        val start = (position % capacity).toInt() * channels
        val count = frames * channels
        val first = min(count, samples.size - start)
        System.arraycopy(source, sourceOffset * channels, samples, start, first)
        if (first < count) {
            System.arraycopy(source, sourceOffset * channels + first, samples, 0, count - first)
        }
    }

    private fun ringRead(position: Long, output: ShortArray, frames: Int) {
        val start = (position % capacity).toInt() * channels
        val count = frames * channels
        val first = min(count, samples.size - start)
        System.arraycopy(samples, start, output, 0, first)
        if (first < count) System.arraycopy(samples, 0, output, first, count - first)
    }

    private fun ringZero(position: Long, frames: Int) {
        val start = (position % capacity).toInt() * channels
        val count = frames * channels
        val first = min(count, samples.size - start)
        samples.fill(0, start, start + first)
        if (first < count) samples.fill(0, 0, count - first)
    }

    private fun multiplyOrMax(left: Long, right: Long): Long = try {
        Math.multiplyExact(left, right)
    } catch (_: ArithmeticException) {
        Long.MAX_VALUE
    }

    private fun addOrMax(left: Long, right: Long): Long = try {
        Math.addExact(left, right)
    } catch (_: ArithmeticException) {
        Long.MAX_VALUE
    }

    private companion object {
        const val PCM_BYTES_PER_SAMPLE = 2L
        const val MICROS_PER_SECOND = 1_000_000L
    }
}

File: ios/Sources/MoQKit/Subscribe/internal/pipeline/PcmRing.swift
import Foundation

struct PcmRingPolicy: Sendable {
    var maxBytes: UInt64 = 64 * 1024 * 1024
    var maxFrames: Int = .max
    var maxDurationUs: UInt64

    init(
        maxBytes: UInt64 = 64 * 1024 * 1024,
        maxFrames: Int = .max,
        maxDurationUs: UInt64
    ) {
        precondition(maxBytes > 0)
        precondition(maxFrames > 0)
        precondition(maxDurationUs > 0)
        self.maxBytes = maxBytes
        self.maxFrames = maxFrames
        self.maxDurationUs = maxDurationUs
    }
}

struct PcmWriteResult: Equatable {
    let acceptedFrames: Int
    var rejectedOldFrames: Int = 0
    var evictedFrames: Int = 0
    var silenceFrames: Int = 0
}

/// Timestamp-positioned Float32 PCM ring (non-interleaved, per channel) with explicit
/// bounded-admission effects.
struct PcmRing {
    private var buffer: [[Float32]]
    private var writeIndex: Int = 0
    private var readIndex: Int = 0

    let rate: Int
    let channels: Int
    private(set) var stalled: Bool = true

    init(rate: Int, channels: Int, policy: PcmRingPolicy) {
        precondition(channels > 0, "invalid channels")
        precondition(rate > 0, "invalid sample rate")
        let durationFrames = Int(ceil(
            Double(rate) * Double(policy.maxDurationUs) / 1_000_000
        ))
        let byteFrames = Int(clamping: policy.maxBytes / UInt64(
            channels * MemoryLayout<Float32>.size
        ))
        let samples = max(1, min(policy.maxFrames, durationFrames, byteFrames))
        precondition(samples > 0, "empty buffer")

        self.rate = rate
        self.channels = channels
        self.buffer = (0..<channels).map { _ in [Float32](repeating: 0, count: samples) }
    }

    /// Timestamp of the current read position in microseconds.
    var timestampUs: UInt64 {
        UInt64(Double(readIndex) / Double(rate) * 1_000_000)
    }

    /// Number of samples available to read.
    var length: Int {
        writeIndex - readIndex
    }

    /// Total capacity in samples.
    var capacity: Int {
        buffer[0].count
    }

    /// Current fill level in milliseconds.
    var fillMs: Double {
        Double(length) / Double(rate) * 1000.0
    }

    // MARK: - Bulk copy helpers

    /// Copy `count` samples from `src` into ring buffer `channel` starting at logical `pos`.
    /// Splits at the wrap point — at most 2 memcpy operations.
    private mutating func ringCopy(channel: Int, pos: Int, src: UnsafePointer<Float32>, count: Int)
    {
        let cap = buffer[channel].count
        let start = pos % cap
        let firstChunk = min(count, cap - start)

        buffer[channel].withUnsafeMutableBufferPointer { dst in
            dst.baseAddress!.advanced(by: start)
                .update(from: src, count: firstChunk)
            if firstChunk < count {
                dst.baseAddress!
                    .update(from: src.advanced(by: firstChunk), count: count - firstChunk)
            }
        }
    }

    /// Copy `count` samples from ring buffer `channel` starting at logical `pos` into `dst`.
    private func ringRead(channel: Int, pos: Int, dst: UnsafeMutablePointer<Float32>, count: Int) {
        let cap = buffer[channel].count
        let start = pos % cap
        let firstChunk = min(count, cap - start)

        buffer[channel].withUnsafeBufferPointer { src in
            dst.update(from: src.baseAddress!.advanced(by: start), count: firstChunk)
            if firstChunk < count {
                dst.advanced(by: firstChunk)
                    .update(from: src.baseAddress!, count: count - firstChunk)
            }
        }
    }

    /// Zero-fill `count` samples in ring buffer `channel` starting at logical `pos`.
    private mutating func ringZero(channel: Int, pos: Int, count: Int) {
        let cap = buffer[channel].count
        let start = pos % cap
        let firstChunk = min(count, cap - start)

        buffer[channel].withUnsafeMutableBufferPointer { dst in
            dst.baseAddress!.advanced(by: start)
                .initialize(repeating: 0, count: firstChunk)
            if firstChunk < count {
                dst.baseAddress!
                    .initialize(repeating: 0, count: count - firstChunk)
            }
        }
    }

    // MARK: - Write / Read / Resize / Reset

    /// Resize the buffer to match a new latency target. Preserves the most recent samples.
    /// Triggers a stall to refill.
    mutating func resize(policy: PcmRingPolicy) {
        let durationFrames = Int(ceil(
            Double(rate) * Double(policy.maxDurationUs) / 1_000_000
        ))
        let byteFrames = Int(clamping: policy.maxBytes / UInt64(
            channels * MemoryLayout<Float32>.size
        ))
        let newCapacity = max(1, min(policy.maxFrames, durationFrames, byteFrames))
        guard newCapacity != capacity else { return }
        precondition(newCapacity > 0, "empty buffer")

        var newBuffer = (0..<channels).map { _ in [Float32](repeating: 0, count: newCapacity) }

        let samplesToKeep = min(length, newCapacity)
        if samplesToKeep > 0 {
            let copyStart = writeIndex - samplesToKeep
            for channel in 0..<channels {
                newBuffer[channel].withUnsafeMutableBufferPointer { dst in
                    ringRead(
                        channel: channel, pos: copyStart, dst: dst.baseAddress!,
                        count: samplesToKeep)
                }
            }
        }

        buffer = newBuffer
        readIndex = writeIndex - samplesToKeep
        stalled = true
    }

    /// Write samples at the position determined by the timestamp.
    /// Accepts raw channel pointers directly from `AVAudioPCMBuffer.floatChannelData`.
    /// Returns explicit admission effects for diagnostics and statistics.
    @discardableResult
    mutating func write(
        timestampUs: UInt64,
        channelData: UnsafePointer<UnsafeMutablePointer<Float32>>,
        frameCount: Int
    ) -> PcmWriteResult {
        var start = Int(round(Double(timestampUs) / 1_000_000.0 * Double(rate)))
        var samples = frameCount
        var discarded = 0
        var silenceFrames = 0

        // First write after init/reset: anchor indices to the incoming timestamp
        // so we don't treat the gap from 0 as overflow.
        if stalled && writeIndex == 0 && readIndex == 0 {
            readIndex = start
            writeIndex = start
        }

        // Ignore samples that are too old (before the read index)
        let offset = readIndex - start
        if offset > samples {
            return PcmWriteResult(acceptedFrames: 0, rejectedOldFrames: samples)
        } else if offset > 0 {
            discarded += offset
            samples -= offset
            start += offset
        }

        let end = start + samples

        // Check if we need to discard old samples to prevent overflow
        let overflow = end - readIndex - buffer[0].count
        if overflow >= 0 {
            stalled = false
            discarded += overflow
            readIndex += overflow
        }

        // Fill gaps with zeros if there's a discontinuity
        if start > writeIndex {
            let gapSize = min(start - writeIndex, buffer[0].count)
            silenceFrames = gapSize
            for channel in 0..<channels {
                ringZero(channel: channel, pos: writeIndex, count: gapSize)
            }
        }

        // Write the actual samples
        let srcOffset = frameCount - samples
        for channel in 0..<channels {
            let src = UnsafePointer(channelData[channel])
            ringCopy(channel: channel, pos: start, src: src.advanced(by: srcOffset), count: samples)
        }

        if end > writeIndex {
            writeIndex = end
        }

        return PcmWriteResult(
            acceptedFrames: samples,
            rejectedOldFrames: max(0, discarded - max(0, overflow)),
            evictedFrames: max(0, overflow),
            silenceFrames: silenceFrames
        )
    }

    /// Reset all state, clearing the buffer and re-entering stalled mode.
    mutating func reset() {
        writeIndex = 0
        readIndex = 0
        stalled = true
        for channel in 0..<channels {
            buffer[channel].withUnsafeMutableBufferPointer { buf in
                buf.baseAddress!.initialize(repeating: 0, count: buf.count)
            }
        }
    }

    /// Read up to `frameCount` samples into the provided output arrays. Returns how many samples were read.
    /// Returns 0 while stalled.
    mutating func read(into output: inout [[Float32]], frameCount: Int) -> Int {
        precondition(output.count == channels, "wrong number of channels")
        guard !stalled else { return 0 }

        let samples = min(writeIndex - readIndex, frameCount)
        guard samples > 0 else { return 0 }

        for channel in 0..<channels {
            precondition(output[channel].count >= samples, "output buffer too small")
            output[channel].withUnsafeMutableBufferPointer { dst in
                ringRead(channel: channel, pos: readIndex, dst: dst.baseAddress!, count: samples)
            }
        }

        readIndex += samples
        return samples
    }
}
